### PR TITLE
Add variable truth tutorial

### LINKED ISSUE #89: Add notebook tutorials for newly added truth/light-curve info
This is a sub-issue of #88.

### PR DESCRIPTION
**Pull request by yymao**
This PR adds the variable truth tutorial and fixes #89. 

Feedback welcome.  There's a few cells that need to be removed before merging; I left them there for now so that the reviewer can run this notebook on NERSC more easily (you will need to use the `u/yymao/pr587-review` branch of `gcr-catalogs` for this notebook to work).

### CODE REVIEW

**Comment by yymao** [comment]
I have updated this tutorial notebook to match the latest truth catalogs. It's now ready for review @JoanneBogart @wmwv

File: notebooks/dc2/truth_variable_info.ipynb [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+    "\n",
+    "**Authors**: Yao-Yuan Mao (@yymao)\n",
+    "\n",
+    "This notebook will illustrate the basics of accessing the variable truth information, such as the light curves for stars and SNe.\n",
+    "\n",
+    "**Prerequisite (optional)**: The Object Table tutorial provides an introduction to `GCRCatalogs`. It is not strictly needed in this tutorial, but if you have questions about the use of `GCRCatalogs`, please check out the Object Table tutorial.\n",
+    "\n",
+    "**Learning objectives**: After going through this notebook, you should be able to:\n",
+    "  1. Know the difference of the various truth catalogs related variable truth information.\n",
+    "  2. Being able to load the light curves for any SNe or stars that are of interset. "
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:17]: omit "being able to".   For 2. you just need "Load the light curves ..."

File: notebooks/dc2/truth_variable_info.ipynb [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+    "GCRCatalogs.set_root_dir(\"/global/cfs/cdirs/lsst/gsharing\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Supernova light curves\n",
+    "\n",
+    "We will start with finding light curves for supernovae.  When accessing DC2 variable truth information, one important concept is that the light curves are stored separetely from a \"summary table\". \n",
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:118]: separetely --> separately

File: notebooks/dc2/truth_variable_info.ipynb [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+    "\n",
+    "In the summary table, each row corresponds to one object (supernova), and it contains \"summary\" truth information such as coordinates, redshift, and time-averaged fluxes, but not the light curves. \n",
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:120]: Maybe add a sentence like
"Here _summary_ implies associated with the source generally, not with a particular observation of the source." 
or modify the sentence you have:

In the summary table, each row corresponds to one object (here a supernova), and it contains _summary_ truth information such as coordinates, redshift, and time-averaged fluxes - that is, attributes associated with the object overall, not with a particular observation of it.  Hence it does not include light curve values.

File: notebooks/dc2/truth_variable_info.ipynb [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+   "outputs": [],
+   "source": [
+    "sorted(sn_summary_cat.list_all_quantities())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will now actually load the data into memory. For this tutorial, we will only need the ID (`id`), coordinates (`ra`, `dec`), redshift. \n",
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:151]: Replace comma before the word "redshift" with the word "and"

File: notebooks/dc2/truth_variable_info.ipynb [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+   "source": [
+    "d_summary = pd.DataFrame(sn_summary_cat.get_quantities([\"id\", \"ra\", \"dec\", \"redshift\"]))\n",
+    "d_summary.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now that we have the summary table, we can use it to select, say supernovae near a certain coordinate or in a certain redshift range. In this tutorial, we will find all supernovae that are within 2 arcmin of (RA, Dec) = (60.5, -36.6) and within z < 0.8. We will store these supernovae in `objects_to_plot` which we will use later to extract and plot the their light curves."
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:170]: Suggest ending sentence with
"all supernovae that are within 2 arcmin of (RA, Dec) = (60.5, -36.6) and satisfy z < 0.8."

File: notebooks/dc2/truth_variable_info.ipynb [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+   "source": [
+    "sn_light_curve_cat = GCRCatalogs.load_catalog(\"desc_dc2_run2.2i_truth_sn_variability\")\n",
+    "sn_light_curve_cat.list_all_quantities()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Unlike the summary table, in the variability table, each row corresponds to one \"measurement\" the flux of a given object, at a given time, with a given filter. Hence, a single supernovae may correspond to many rows. \n",
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:211]: 1. Should not be a comma after variability table.  2. There is a missing word ("of") or missing punctuation (colon) after "measurement".   3. "corresponds" is maybe not the best word here.   I suggest

Unlike the summary table, in the variability table each row corresponds to one "measurement": the flux of a given object, at a given time, with a given filter. Hence, a single supernova is usually represented by many rows.

File: notebooks/dc2/truth_variable_info.ipynb [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+   "source": [
+    "d_lc = pd.DataFrame(first(sn_light_curve_cat.get_quantities(sn_light_curve_cat.list_all_quantities(), return_iterator=True)))\n",
+    "d_lc.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can see that from above that the first five rows are all different measurements (at different time or with different filters) for a single object (`MS_9177_2005`). \n",
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:230]: Suggest
You can see from the above the the first five rows are all different measurements (at different times and not always with the same filter) for a single...

File: notebooks/dc2/truth_variable_info.ipynb [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+    "\n",
+    "Now that we have a basic idea of how the variability table is structured, we can start to search for the light curves for the supernovae that we identified earlier. In short, we need to extract all rows that have IDs overlapping with the IDs in `objects_to_plot`. \n",
+    "\n",
+    "Because the variability table is a large table, we don't want to load the whole thing and then do the search. By specifying `filters` in `get_quantities`, the backend code will do the search on each smaller chunk, which will save the runtime memory. \n",
+    "\n",
+    "The filter (i.e., search query) we will use here looks a bit complicated, but it really just means find all rows whose `id` is in `objects_to_plot[\"id\"].values`: \n",
+    "```python\n",
+    "GCRQuery((lambda x: np.isin(x, objects_to_plot[\"id\"].values), \"id\"))\n",
+    "```\n",
+    "Even with this filter, the cell below will take a while to run. If runtime is a concern, you may want to use the SQL database that we also provide to access these information instead."
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:240]: Change "these information" to "this information"

File: notebooks/dc2/truth_variable_info.ipynb [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+    "    ax_this.set_ylim(25, 10)\n",
+    "\n",
+    "fig.tight_layout()        "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we have the light curves! You might notice that we used `delta_flux` to get the magnidue. Here `delta_flux` is the flux difference to the inifinite time averaged flux value. For SNe, the infinite time-averaged fluxes are always zero, and hence `delta_flux` would just store the actual flux values. As we will see later, for stars we will have to add the infinite time-averaged fluxes to `delta_flux` to get actual flux values."
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:298]: magnidue -->.  magnitude

File: notebooks/dc2/truth_variable_info.ipynb [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    "sorted(star_summary_cat.list_all_quantities())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Here the fluxes (`flux_*`) in the stars' summary table are time-averaged fluxes over *infinite* time, which we will need when we plot the light curves.\n",
+    "\n",
+    "Also, not all stars in this summary table are \"variable\"; some are rather static. The column `above_threshold` tells us if the star is \"variable enough\" to be included in the variability table. So here we will also grab it when we are looking for interseting variable stars. "
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:336]: interseting  -->. interesting

File: notebooks/dc2/truth_variable_info.ipynb [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+   "outputs": [],
+   "source": [
+    "d_summary = pd.DataFrame(star_summary_cat.get_quantities([\"id\", \"ra\", \"dec\", \"above_threshold\"] + [f\"flux_{band}\" for band in \"ugrizy\"]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Another important difference between stars and supernovae is that for stars, their infinite time-averaged fluxes are usualy non-zero! We can verify that:"
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:352]: usualy --> usually

File: notebooks/dc2/truth_variable_info.ipynb [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+   "outputs": [],
+   "source": [
+    "star_lc_cat = GCRCatalogs.load_catalog(\"desc_dc2_run2.2i_truth_star_variability\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Again, recall that in the \"variability table\", each row corresponds to one measurement, and we need to collect all the rows that have `id` values overlap with `objects_to_plot[\"id\"].values`. \n",
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:407]: overlap    --> overlapping

File: notebooks/dc2/truth_variable_info.ipynb [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+    "\n",
+    "Also remember that the variability table is a large table (even larger in the star case). We will use the same `filters` trick again to reduce the memory footprint, but the cell below will still result in ~4 GB at peak memory usage. It will take a while to run. If runtime is a concern, you may want to use the SQL database that we also provide to access these information instead."
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:409]: these information  --> this information

File: notebooks/dc2/truth_variable_info.ipynb [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+    "    filters=[GCRQuery((lambda x: np.isin(x, objects_to_plot[\"id\"].values), \"id\"))],\n",
+    "))\n",
+    "d_lc.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Once we collect the light curve infomation in `d_lc`. We can use the same `groupby` trick to split the data frame by ID and bandpass, and proceed to make light curve plot."
```
Comment (JoanneBogart) [inline notebooks/dc2/truth_variable_info.ipynb:429]: Add "a" or "the" before "light curve plot".    E.g.,
..and proceed to make the light curve plot

**Review by JoanneBogart — CHANGES_REQUESTED** [review]
Generally looks good!  I've made a number of minor suggestions - fixing typos, making the text more idiomatic or, in one or two cases, trying for more clarity.  Please also consider whether there is a way to make the variability in the  first star plot more evident.

**Comment by yymao** [comment]
@JoanneBogart thanks for the review. I've addressed all the comments. You can see the detail changed in commit e8e087d. Regarding making the variability in the first star plot more evident: the second variability plot (difference in magnitude) is meant to address the issue. So instead of changing the plot, I just updated the text to make this point more explicit. 

I've also tested the truth_match notebook and everything works fine. Numpy is deprecating the use `np.int` in favor of `np.int64` so I made that change in cba6c3d.

Ready for re-review.

**Review by JoanneBogart — APPROVED** [review]
LGTM.

**Pull request merged**